topdown/eval.go

```diff
@@ -2516,16 +2560,23 @@ func (e evalVirtualComplete) evalValue(iter unifyIterator) error {
 
 	var prev *ast.Term
 
-	for i := range e.ir.Rules {
-		next, err := e.evalValueRule(iter, e.ir.Rules[i], prev)
+	for _, rule := range e.ir.Rules {
+		next, err := e.evalValueRule(iter, rule, prev)
 		if err != nil {
-			return err
+			// we ignore EE signals if we cannot exit early ourselves
+			_, ok := err.(*signalEarlyExit)
+			if !ok || e.ir.EarlyExit {
```

Comment:
nit: What happens if `ok && !e.ir.EarlyExit`? I don't think this is possible but that makes me wonder why we're checking for `e.ir.EarlyExit` in the first place.

Reply:
Might be easier to follow if not condensed, like this:
```go
_, ok := err.(*signalEarlyExit)
if ok && e.ir.EarlyExit {
  return err // signal EE to parent
}
if !ok {
  return err // some other error
}
```
So, `ok && !e.ir.EarlyExit` would be the case for

```rego
p { q }

q = x { x := true; trace("a") } # can't EE because of x
q = x { x := false; trace("b") }
```
- the `evalVirtualComplete.evalValue` for `p` will call
  - the `evalVirtualComplete.evalValue` for `q`'s first body, which will call
    - the iterator passed along that'll return the EE signal error
  - and check the error returned: if it's an EE signal, it has to ignore it and keep evaluating its other rule bodies

So if the code read
```go
if ok {
  return err // signal EE to parent
}
if !ok {
  return err // some other error
}
```
we wouldn't see the conflict error that we should see for `q`.

And if it was only
```go
if !ok {
  return err // some other error
}
```
we'd never EE, even if we can.